pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,443 @@
+		return err
+	}
+	if activeGW.Name == c.nodeConfig.Name {
+		klog.V(2).InfoS("Adding/updating flows to remote Gateway Node for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+		localGatewayIP := net.ParseIP(activeGW.GatewayIP)
+		if err := c.ofClient.InstallMulticlusterGatewayFlows(
+			activeGW.Name+ciImport.Name,
+			peerConfigs,
+			tunnelPeerIP,
+			localGatewayIP); err != nil {
+			return fmt.Errorf("failed to install flows to remote Gateway in ClusterInfoImport %s: %v", ciImport.Name, err)
+		}
+	} else {
+		klog.V(2).InfoS("Adding/updating flows to the local active Gateway for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+		tunnelPeerIP := net.ParseIP(activeGW.InternalIP)
+		if err := c.ofClient.InstallMulticlusterNodeFlows(
+			activeGW.Name+ciImport.Name,
+			peerConfigs,
+			tunnelPeerIP); err != nil {
+			return fmt.Errorf("failed to install flows to Gateway %s: %v", activeGW.Name, err)
+		}
+	}
+
+	c.installedCIImports.Add(ciImport)
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForCIImport(ciImp *mcv1alpha1.ClusterInfoImport, activeGWName string) error {
+	key := activeGWName + ciImp.Name
+	if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+		return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway Node %s: %v", ciImp.Name, err)
+	}
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForGateway(gwName string) error {
+	cachedciImps := c.installedCIImports.List()
+	for _, ciImp := range cachedciImps {
+		ciImport := ciImp.(*mcv1alpha1.ClusterInfoImport)
+		key := gwName + ciImport.Name
+		if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+			return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway in ClusterInfoImport %s: %v",
+				ciImport.Name, err)
+		}
+		c.installedCIImports.Delete(ciImport)
+	}
+	return nil
+}
+
+// getActiveGateway compares Gateway's CreationTimestamp to get the active Gateway,
+// The last created Gateway will be the active Gateway.
+func (c *MCRouteController) getActiveGateway() (*mcv1alpha1.Gateway, error) {
+	gws, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	if len(gws) == 0 {
+		return nil, nil
+	}
+	// Comparing Gateway's CreationTimestamp to get the last created Gateway.
+	lastCreatedGW := gws[0]
+	for _, gw := range gws {
+		if lastCreatedGW.CreationTimestamp.Before(&gw.CreationTimestamp) {
+			lastCreatedGW = gw
+		}
+	}
+	return lastCreatedGW, nil
+}
+
+func generatePeerConfigs(subnets []string, gatewayIP net.IP) (map[*net.IPNet]net.IP, error) {
+	peerConfigs := make(map[*net.IPNet]net.IP, len(subnets))
+	for _, subnet := range subnets {
+		_, peerCIDR, err := net.ParseCIDR(subnet)
+		if err != nil {
+			klog.Errorf("Failed to parse subnet %s", subnet)
```

Comment:
structured logging, but I think this log is not needed as the error has been logged on callers multiple times.
normally, to avoid duplicate logs, a function should either wrap the error with current context and pop up it without logging it or swallow the error but log it if it's not supposed to pop up it further, so it's guaranteed that an error will be logged exactly once with full context.